### PR TITLE
Check for unbounded recursion during dropck

### LINKED ISSUE #22443: Infinite loop/stack overflow in rustc (dropck) on non-regular data type
The compiler goes into an infinite loop compiling this code (an incomplete finger tree implementation):

```rust
struct Digit<T> {
    elem: T,
    next: Option<Box<Digit<T>>>
}

enum Node<T> {
    Node2(T, T),
    Node3(T, T, T)
}

enum FingerTree<T> {
    Empty,
    Single(T),
    Deep(Digit<T>, Box<FingerTree<Node<T>>>, Digit<T>)
}

impl <T> FingerTree<T> {
    fn dequeue(&self, elem: T) -> FingerTree<T> {
        match self {
            Empty => FingerTree::Single(elem)
        }
    }
}

fn main() {
    let ft = FingerTree::Empty;
    let ft2 = ft.dequeue(123);
}
```

Tested with latest nightly. I can also cause this to happen by pasting this into the box on the rust-lang homepage.

### PR DESCRIPTION
**Pull request by pnkfelix**
Check for unbounded recursion during dropck.

Such recursion can be introduced by the erroneous use of non-regular types (aka types employing polymorphic recursion), which Rust does not support.

Fix #22443

### CODE REVIEW

File: src/test/compile-fail/dropck_no_diverge_on_nonregular_2.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Issue 22443: Reject code using non-regular types that would
+// otherwise cause dropck to loop infinitely.
```
Comment (nikomatsakis) [inline src/test/compile-fail/dropck_no_diverge_on_nonregular_2.rs]: what is the difference between this and the first test?

Comment (pnkfelix) [inline src/test/compile-fail/dropck_no_diverge_on_nonregular_2.rs]: the comment below says why.  (The bug report observed different behaviors, both bad, depending on the order of Digit  versus Box)

Comment (nikomatsakis) [inline src/test/compile-fail/dropck_no_diverge_on_nonregular_2.rs]: I see, I think it'd be helpful to write in the comment something like "In contrast to Digit after Box, which is the other test."

Comment (pnkfelix) [inline src/test/compile-fail/dropck_no_diverge_on_nonregular_2.rs]: Okay, will do

**Comment by nikomatsakis** [comment]
r+ modulo nit

**Comment by pnkfelix** [comment]
(Note for rollup builders: I marked this as rollup, but do note that it adds a new error code, which increases its chances for collision with another PR that *also* adds the same error code, especially since the same error code might be added in a different file.)

**Comment by Manishearth** [comment]
This caused a slew of dropck overflows in the rollup.

a sample:

```
/opt/rust/src/libsyntax/fold.rs:992:17: 992:56 error: overflow while adding drop-check rules for util::small_vector::IntoIter<ast::ImplItem> [E0320]
/opt/rust/src/libsyntax/fold.rs:992                 folder.fold_impl_item(item).into_iter()
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rust/src/libsyntax/fold.rs:992:17: 992:56 note: overflowed on struct ast::StructField_ field kind type: ast::StructFieldKind
/opt/rust/src/libsyntax/fold.rs:992                 folder.fold_impl_item(item).into_iter()
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rust/src/libsyntax/fold.rs:992:17: 992:44 error: overflow while adding drop-check rules for util::small_vector::SmallVector<ast::ImplItem> [E0320]
/opt/rust/src/libsyntax/fold.rs:992                 folder.fold_impl_item(item).into_iter()
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rust/src/libsyntax/fold.rs:992:17: 992:44 note: overflowed on struct ast::StructField_ field kind type: ast::StructFieldKind
/opt/rust/src/libsyntax/fold.rs:992                 folder.fold_impl_item(item).into_iter()
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rust/src/libsyntax/fold.rs:1028:13: 1029:44 error: overflow while adding drop-check rules for core::iter::Map<util::small_vector::IntoIter<ptr::P<ast::Method>>, [closure /opt/rust/src/libsyntax/fold.rs:1029:22: 1029:43]> [E0320]
/opt/rust/src/libsyntax/fold.rs:1028             folder.fold_method(method).into_iter()
/opt/rust/src/libsyntax/fold.rs:1029                 .map(|m| ProvidedMethod(m)).collect()
/opt/rust/src/libsyntax/fold.rs:1028:13: 1029:44 note: overflowed on struct ast::StructField_ field kind type: ast::StructFieldKind
/opt/rust/src/libsyntax/fold.rs:1028             folder.fold_method(method).into_iter()
/opt/rust/src/libsyntax/fold.rs:1029                 .map(|m| ProvidedMethod(m)).collect()
```

**Comment by Manishearth** [comment]
[Full log](https://manishearth.pastebin.mozilla.org/8823560)

Looks like they were caused during `make check` for various crates

**Comment by pnkfelix** [comment]
weird!  Okay, I'll look into that.

**Comment by pnkfelix** [comment]
@Manishearth wow, I totally failed to `make rustc-stage2` (and apparently nothing in `run-pass` has a structure that exercises something like this...) ; sorry about that!

**Comment by Manishearth** [comment]
Hah. I don't think anyone expects rustc-stage2 from anything but large PRs.

So, this PR needs rustc-stage2 to be tested properly? I don't envy your next few days :grimacing:  (I had to do the same for `-Z extra-plugins`)

**Comment by pnkfelix** [comment]
@Manishearth no, now that I know about it, I can make a test applicable for the stage1 compiler (and put it into `run-pass`).  (Also, `make check-stage1` *would* have could this; but not `check-stage1-rpass`)

**Comment by Manishearth** [comment]
That sounds less painful :smiley:

**Comment by pnkfelix** [comment]
Reduced test case:

```rust
#![allow(non_camel_case_types)]

pub fn noop_fold_impl_item() -> SmallVector<ImplItem> {
    loop  { }
}

pub struct SmallVector<T>(P<T>);
pub struct ImplItem(P<S01_Method>);

struct P<T>(Box<T>);

struct S01_Method(P<S02_Generics>);
struct S02_Generics(P<S03_TyParam>);
struct S03_TyParam(P<S04_TyParamBound>);
struct S04_TyParamBound(S05_PolyTraitRef);
struct S05_PolyTraitRef(S06_TraitRef);
struct S06_TraitRef(S07_Path);
struct S07_Path(Vec<S08_PathSegment>);
struct S08_PathSegment(S09_PathParameters);
struct S09_PathParameters(P<S10_ParenthesizedParameterData>);
struct S10_ParenthesizedParameterData(Option<P<S11_Ty>>);
struct S11_Ty(P<S12_Expr>);
struct S12_Expr(P<S13_Block>);
struct S13_Block(Vec<P<S14_Stmt>>);
struct S14_Stmt(P<S15_Decl>);
struct S15_Decl(P<S16_Local>);
struct S16_Local(P<S17_Pat>);
struct S17_Pat(P<S18_Mac>);
struct S18_Mac(Vec<P<S19_TokenTree>>);
struct S19_TokenTree(P<S20_Token>);
struct S20_Token(P<S21_Nonterminal>);
struct S21_Nonterminal(P<S22_Item>);
struct S22_Item(P<S23_EnumDef>);
struct S23_EnumDef(Vec<P<S24_Variant>>);
struct S24_Variant(P<S25_VariantKind>);
struct S25_VariantKind(P<S26_StructDef>);
struct S26_StructDef(Vec<P<S27_StructField>>);
struct S27_StructField(P<S28_StructFieldKind>);
struct S28_StructFieldKind;

pub fn main() {}
```

**Comment by pnkfelix** [comment]
So, after walking through the debug log, I see that a likely problem here is that I am incrementing the recursion counter even for the recursion that descends from `struct A(B);` to the definition of `struct B { ... }`.

That kind of recursive descent is just standard traversal over the inductive structure of the type; i.e., it cannot lead to infinite loops.

The only kind of traversal that can lead to infinite loops are those that go through references, i.e. cases like `&B`, `Box<B>`, `Vec<B>`, ....; such references fall into two categories: non-owning references and owning references.  Non-owning references like `&B` are ignored by dropck.  Owning references are captured by `Box<B>` and `PhantomData<B>` (and in the future, `Box<B>` should just be another instance of `PhantomData<B>`.

SO: For the short term, my plan is to keep a separate counter that is incremented only when the recursion traverses through an owning reference, as outlined above.

-----

For the long term, however, I am a little worried that such an approach does not scale up properly.  I.e. one is still going to be able to make programs (probably *real* programs) that will exceed the threshold, and I am not convinced the solution of "just tell such crates to increase the value of the recursion counter" will be satisfying in practice.

Instead of just using a fixed recursion limit, we may want to instead make the recursion threshold a *function* of the number of nodes in the AST, or some similar value that can be easily derived from the static text.  E.g. we instead made the threshold where we signal an error be equal to `#nodes * recursion_limit` (where we of course would use saturated multiply there, and perhaps even 64-bit arithmetic for it).

Anyway, that's just a stray thought.  For this PR, I'm certainly going to take the short-cut of just not increasing this counter so fast.

**Comment by pnkfelix** [comment]
hmm, my simplistic attempt to apply the strategy from the previous comment does not seem to actually catch the test from #22443

**Comment by pnkfelix** [comment]
(Note again for rollup builders: I marked this as rollup, but do note that it adds a new error code, which increases its chances for collision with another PR that also adds the same error code, especially since the same error code might be added in a different file.  On the bright side, I did run `make check-stage2` ahead of time this time. :)  )

**Pull request merged**